### PR TITLE
Slight crafting refactors, fixes trash cannon not being able to fire and some foods exploding

### PR DESCRIPTION
**Pull request by Maia-J**
## About The Pull Request

Introduces `recipe_flags` instead of `delete_contents` with 2 flags `RECIPE_KEEP_CONTENTS` and `RECIPE_DONT_TRANSFER_REAGENTS`.

## Why It's Good For The Game

Fixes the trash cannon being crafted full of tar and makes the code prettier

## Changelog

:cl:
fix: Crafting the trash cannon no longer fills it with tar, allowing it to be actually fired again!
fix: Rootbread no longer takes on reagents from its crafting recipe, preventing peanut butter and banana rootwich explosions
/:cl:

No other CL because the rest isn't player facing

### CODE REVIEW

**Comment by SyncIt21** [comment]
Don't like how hacky and unessassary this is.

If the problem is with just 1 crafting recipe then override this proc for the trash cannon crafting recipe 
https://github.com/tgstation/tgstation/blob/976508161921e7a62ef71c0dc331030be060ed14/code/datums/components/crafting/_recipes.dm#L94
And clear the reagents transferred there.

**Comment by Majkl-J** [comment]
> Don't like how hacky this is.
> 
> If the problem is with just 1 crafting recipe then override this proc for the trash cannon crafting recipe
> 
> https://github.com/tgstation/tgstation/blob/976508161921e7a62ef71c0dc331030be060ed14/code/datums/components/crafting/_recipes.dm#L94
> 
> 
> And clear the reagents transferred there.
> No need to introduce new flags just to create exceptions for each crafting recipe

I feel like it is worth it for the future support. The variable delete_contents already exists for just one item and was added for exactly the same reason: To make it easier to add new recipes with the functionality

**Comment by SyncIt21** [comment]
delete_contents does it job cleanly but your don't transfer reagents flag is inconsistent in its behavior. 

That is you Don't transfer the reagents but you still clear out the original reagents? Clearing and transferring is one atomic operation, you can't split it apart with bit flags just for one recipe sake,  That causes confusion.

Solution is to simply not have flags like this and actually use the dedicated proc for specialized recipes like these.

**Comment by Majkl-J** [comment]
> delete_contents does its job cleanly but your don't transfer reagents flag is inconsistent in its behavior.
> 
> That is you Don't transfer the reagents but you still clear out the original reagents? Clearing and transferring is one atomic operation, you can't split it apart with bit flags just for one recipe sake, That causes confusion.
> 
> Solution is to simply not have flags and actually use the dedicated proc for specialized recipes like these.

Alright that's an oversight on my part that I missed. Time to fix

File: code/modules/food_and_drinks/recipes/tablecraft/recipes_misc.dm [hunk 1]
```diff
@@ -426,7 +426,7 @@
 	)
 	result = /obj/item/storage/fancy/pickles_jar
 	category = CAT_MISCFOOD
-	delete_contents = FALSE
+	recipe_flags = 0
```
Comment (Ghommie) [inline code/modules/food_and_drinks/recipes/tablecraft/recipes_misc.dm:429]:
```suggestion
	recipe_flags = NONE
```

**Comment by Majkl-J** [comment]
Turns out there's significantly more issues with how this works holy shit

**Comment by SyncIt21** [comment]
> Turns out there's significantly more issues with how this works holy shit
> 
> Update: Nvm I was just dumb. Ready for review again

You are over complicating this. Your whole PR can be just 2 lines of code. Remove all changes you made and just do this

/datum/crafting_recipe/trash_cannon/on_craft_completion(mob/user, atom/result) 
  result.reagents.clear_reagents()

No need for fancy bit flags

**Comment by Majkl-J** [comment]
> > Turns out there's significantly more issues with how this works holy shit
> > Update: Nvm I was just dumb. Ready for review again
> 
> You are over complicating this. Your whole PR can be just 2 lines of code. Remove all changes you made and just do this
> 
> ```
> /datum/crafting_recipe/trash_cannon/on_craft_completion(mob/user, atom/result) 
>     result.reagents.clear_reagents()
> ```
> 
> No need for fancy bit flags

After talking with several people, turns out there are several recipes with funny consequences as a result of transferring all reagents. Peanut butter and banana rootwiches EXPLODE, wiping their reagents and making them nonedible because rootbread needs water and bananas usually carry potassium. So yea, this is one way to fix it

**Comment by SyncIt21** [comment]
The correct solution is to change the ingredients required to craft these recipes so that after transfering the reagents they behave like normal.

For example for your trash gun remove requiring tar from the ingredients so no reagents get transferred.

Similarly for other food items check what the initial reagents present in the item are and then set them as it's crafting ingredients.

This way you don't have to change the main crafting code but only the recipes

**Comment by Majkl-J** [comment]
> The correct solution then is to change the ingredients required to craft these recipes so that after transfering the reagents they behave like normal.
> 
> For example for your trash gun remove requiring tar from the ingredients so no reagents get transferred.
> 
> Similarly for other food items check what the initial reagents present in the item are and then set them as it's crafting ingredients.
> 
> This way you don't have to change the main crafting code but only the recipes

That feels like just hiding the issue at hand instead of fixing it

**Comment by SyncIt21** [comment]
> That feels like just hiding the issue at hand instead of fixing it

What exactly is it hiding? Creating exceptions for which items require reagent transfer and which don't is hacky, logically all items should inherit their ingredients cause that makes sense.

**Comment by Majkl-J** [comment]
> What exactly is it hiding? Creating exceptions for which items require reagent transfer and which don't is hacky, logically all items should inherit their ingredients cause that makes sense.
> 
> What you are doing is writing snowflake code for specific items, that's just bad code smell to be honest

What also makes sense is dough needing water to be made, but dough itself isn't exactly full of water when you're done with it. As for maint tar, it is there to make the recipe actually difficult to achieve, because all the other components are plentiful onboard. The reagent transferring itself had to be snowflaked twice by now because there's a lot of items that you don't want the transferring to happen to.

File: code/__DEFINES/recipes.dm [hunk 2]
```diff
@@ -0,0 +1,4 @@
+/// If the recipe deletes contents of the crafted item (Only works with storage items, used for ammo boxes, donut boxes, internals boxes, etc)
+#define RECIPE_DELETE_CONTENTS (1<<0)
```
Comment (Ghommie) [inline code/__DEFINES/recipes.dm:2]: By default, all recipes delete the contents of the resulting object, so this should be a flag for not deleting the contents, which only affects a handful and is easier to maintain.

Comment (Majkl-J) [reply code/__DEFINES/recipes.dm:2]: Makes sense

**Comment by Majkl-J** [comment]
Found another. This stuff is so broken it's honestly silly. Stepped seraka mushrooms keep their toxin, making the recipe utterly useless
image

File: code/modules/food_and_drinks/recipes/tablecraft/recipes_lizard.dm [hunk 3]
```diff
@@ -442,6 +444,7 @@
 	)
 	result = /obj/item/food/steeped_mushrooms
 	category = CAT_LIZARD
+	recipe_flags = RECIPE_KEEP_CONTENTS
```
Comment (Ghommie) [inline code/modules/food_and_drinks/recipes/tablecraft/recipes_lizard.dm:447]: Wrong flag? It's a food item, not a container.

Comment (Majkl-J) [reply code/modules/food_and_drinks/recipes/tablecraft/recipes_lizard.dm]: My bad, did that one really late into the night and it shows

**Comment by Ghommie** [comment]
Ok, so, there's this little issue I've been thinking about after reading the comments. Reagents not being transferred not only means there won't be bad reactions like water + potassium, but that other chems wo't be transferred at all. Logically, if you make the rootdough item with poisonous potatoes, it should be poisonous as well, though this PR impedes that.

I think a proc to handle this sort of reagent problems in place of a generic `holder.trans_to(...)` call would make a better solution, as it lets us fix it case by case without neesssarily having to alter the recipe.

**Comment by SyncIt21** [comment]
> I think a proc to handle this sort of reagent problems in place of a generic holder.trans_to(...) call would make a better solution, as it lets us fix it case by case without having to alter the recipe.

What if we create a proc called `CheckReagents(datum/reagents/holder)`? 
Just like the `CheckParts()` proc defined on the atom level we pass the reagents transferred from the ingredients to this proc and let each individual atom decide what they want to do with them

**Comment by Majkl-J** [comment]
> > I think a proc to handle this sort of reagent problems in place of a generic holder.trans_to(...) call would make a better solution, as it lets us fix it case by case without having to alter the recipe.
> 
> What if we create a proc called `CheckReagents(datum/reagents/holder)`? Just like the `CheckParts()` proc defined on the atom level we pass the reagents transferred from the ingredients to this proc and let each individual atom decide what they want to do with them
> 
> 1)Default behavior would be transfer all reagents 2)Reagents container can override this and do nothing as they are responsible for handling their own reagents 3)Same thing for trash cannon override and do nothing so we don't get tar inside the cannon.
> 
> With this we can implement customized transfering behavior(deciding which reagents get transferred if any and which don't), Rather than a binary yes/no transfer flag.
> 
> Initial I thought why not just pass the reagent holder as one of the parts to `CheckParts()` but we would have to change the behavior of every existing proc to check for this holder in their parts list and transfer the reagents which would be way too cumbersome so a seperate `CheckReagents()` proc is better

That's probably a better idea, I'll swap this thing to draft (Again) and get to working on implementing that instead

**Pull request closed without merging**